File: leetcode/1376/solution.py
"""
https://leetcode.com/problems/time-needed-to-inform-all-employees/description/
1376. Time Needed to Inform All Employees

A company has n employees with a unique ID for each employee from 0 to n - 1. The head of the company is the one with headID.

Each employee has one direct manager given in the manager array where manager[i] is the direct manager of the i-th employee, manager[headID] = -1. Also, it is guaranteed that the subordination relationships have a tree structure.

The head of the company wants to inform all the company employees of an urgent piece of news. He will inform his direct subordinates, and they will inform their subordinates, and so on until all employees know about the urgent news.

The i-th employee needs informTime[i] minutes to inform all of his direct subordinates (i.e., After informTime[i] minutes, all his direct subordinates can start spreading the news).

Return the number of minutes needed to inform all the employees about the urgent news.
"""
from typing import List
from collections import defaultdict


class Solution:
    def numOfMinutes(
        self, n: int, headID: int, manager: List[int], informTime: List[int]
    ) -> int:
        # Create a graph
        graph = defaultdict(list)
        for e, m in enumerate(manager):
            graph[m].append(e)

        # DFS approach
        def dfs(i: int, minutes: int) -> int:
            if len(graph[i]) == 0:
                return minutes
            return max([dfs(j, minutes + informTime[i]) for j in graph[i]])

        return dfs(headID, 0)

        # BFS approach
        # queue = deque()
        # max_value = 0
        # queue.append((headID, 0))
        # while queue:
        #    man, minutes = queue.popleft()
        #    if len(graph[man]) == 0:
        #        max_value = max(max_value, minutes)
        #    else:
        #        for sub in graph[man]:
        #            queue.append((sub, minutes + informTime[man]))

        # return max_value

File: leetcode/1376/solution.rs
#[allow(dead_code)]
struct Solution {}

impl Solution {
    pub fn num_of_minutes(n: i32, head_id: i32, manager: Vec<i32>, inform_time: Vec<i32>) -> i32 {
        // DFS approach
        // Create graph
        let mut graph: Vec<Vec<usize>> = vec![vec![]; n as usize];
        for (e, &m) in manager.iter().enumerate() {
            if m != -1 {
                graph[m as usize].push(e);
            }
        }

        Self::dfs(head_id as usize, 0, &graph, &inform_time)
    }

    fn dfs(m: usize, minutes: i32, graph: &Vec<Vec<usize>>, time: &Vec<i32>) -> i32 {
        let subordinates: &Vec<usize> = &graph[m];
        let mut max_minutes = 0;
        if subordinates.len() == 0 {
            return minutes;
        }
        for &sub in subordinates {
            max_minutes = max_minutes.max(Self::dfs(sub, minutes + time[m], graph, time));
        }
        max_minutes
    }

    // pub fn num_of_minutes(n: i32, head_id: i32, manager: Vec<i32>, inform_time: Vec<i32>) -> i32 {
    // BFS approach
    // use std::collections::{HashMap, VecDeque};

    // // Create graph
    // let mut graph: HashMap<i32, Vec<i32>> = HashMap::new();
    // for (e, &m) in manager.iter().enumerate() {
    // let e = e as i32;
    // graph.entry(m).or_insert(vec![]).push(e);
    // }

    // // enqueue the first item (the head of the company)
    // let mut queue: VecDeque<(i32, i32)> = VecDeque::with_capacity(n as usize);
    // let mut answer = 0;
    // queue.push_back((head_id, 0));
    // while let Some((man, minutes)) = queue.pop_front() {
    // // If manager has subordinate(s), push them to the queue
    // // If not, update answer
    // match graph.get(&man) {
    // Some(subordinates) => {
    // for &sub in subordinates {
    // queue.push_back((sub, minutes + inform_time[man as usize]));
    // }
    // }
    // None => {
    // answer = answer.max(minutes);
    // }
    // }
    // }

    // answer
    // }
}
